pkg/component/apiserver/encryptionconfiguration.go

```diff
@@ -0,0 +1,151 @@
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	secretsutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+var encryptionCodec runtime.Codec
+
+func init() {
+	encryptionScheme := runtime.NewScheme()
+	utilruntime.Must(apiserverconfigv1.AddToScheme(encryptionScheme))
+
+	var (
+		ser = json.NewSerializerWithOptions(json.DefaultMetaFactory, encryptionScheme, encryptionScheme, json.SerializerOptions{
+			Yaml:   true,
+			Pretty: false,
+			Strict: false,
+		})
+		versions = schema.GroupVersions([]schema.GroupVersion{
+			apiserverconfigv1.SchemeGroupVersion,
+		})
+	)
+
+	encryptionCodec = serializer.NewCodecFactory(encryptionScheme).CodecForVersions(ser, ser, versions, versions)
+}
+
+const secretETCDEncryptionConfigurationDataKey = "encryption-configuration.yaml"
+
+// ReconcileSecretETCDEncryptionConfiguration reconciles the ETCD encryption secret configuration.
+func ReconcileSecretETCDEncryptionConfiguration(
+	ctx context.Context,
+	c client.Client,
+	secretsManager secretsmanager.Interface,
+	config ETCDEncryptionConfig,
+	secretETCDEncryptionConfiguration *corev1.Secret,
+	secretNameETCDEncryptionKey string,
+	roleLabel string,
+) error {
+	options := []secretsmanager.GenerateOption{
+		secretsmanager.Persist(),
+		secretsmanager.Rotate(secretsmanager.KeepOld),
+	}
+
+	if config.RotationPhase == gardencorev1beta1.RotationCompleting {
+		options = append(options, secretsmanager.IgnoreOldSecrets())
+	}
+
+	keySecret, err := secretsManager.Generate(ctx, &secretsutils.ETCDEncryptionKeySecretConfig{
+		Name:         secretNameETCDEncryptionKey,
+		SecretLength: 32,
+	}, options...)
+	if err != nil {
+		return err
+	}
+
+	keySecretOld, _ := secretsManager.Get(secretNameETCDEncryptionKey, secretsmanager.Old)
+
+	encryptionConfiguration := &apiserverconfigv1.EncryptionConfiguration{
+		Resources: []apiserverconfigv1.ResourceConfiguration{{
+			Resources: config.Resources,
+			Providers: []apiserverconfigv1.ProviderConfiguration{
+				{
+					AESCBC: &apiserverconfigv1.AESConfiguration{
+						Keys: etcdEncryptionAESKeys(keySecret, keySecretOld, config.EncryptWithCurrentKey),
+					},
+				},
+				{
+					Identity: &apiserverconfigv1.IdentityConfiguration{},
+				},
+			},
+		}},
+	}
+
+	data, err := runtime.Encode(encryptionCodec, encryptionConfiguration)
+	if err != nil {
+		return err
+	}
+
+	secretETCDEncryptionConfiguration.Labels = map[string]string{v1beta1constants.LabelRole: roleLabel}
+	secretETCDEncryptionConfiguration.Data = map[string][]byte{secretETCDEncryptionConfigurationDataKey: data}
+	utilruntime.Must(kubernetesutils.MakeUnique(secretETCDEncryptionConfiguration))
+	desiredLabels := utils.MergeStringMaps(secretETCDEncryptionConfiguration.Labels) // copy
+
+	if err := c.Create(ctx, secretETCDEncryptionConfiguration); err == nil || !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	// reconcile labels of existing secret
```

Comment:
Maybe it is just me, but I got confused here mainly due to the fact that the happy path short-cuts out of this function if create succeeds (`err == nil`), which I overlooked. Not sure if expanding the comment would help.

For example something along the lines of:
```suggestion
	// creation of secret failed as it already exists => reconcile labels of existing secret
```

Reply:
I just moved existing code here, but I'm OK to update the comment